internal/util/traffic_policy_test.go

```diff
@@ -0,0 +1,119 @@
+package util
+
+import (
+	"encoding/json"
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestIsLegacyPolicy(t *testing.T) {
+	testCases := []struct {
+		name             string
+		msg              json.RawMessage
+		expectedIsLegacy bool
+	}{
+		{
+			name:             "json has top-level 'inbound' legacy keys",
+			msg:              []byte(`{"inbound":"some_val", "on_http_request": {"a": "b"}}`),
+			expectedIsLegacy: true,
+		},
+		{
+			name:             "json has top-level 'outbound' legacy keys",
+			msg:              []byte(`{"top_level_key":"some_val","outbound":{"eleven":"twelve"}}`),
+			expectedIsLegacy: true,
+		},
+		{
+			name:             "json only has phase-based naming top-level keys",
+			msg:              []byte(`{"on_tcp_connect":"some_val","on_http_request":{"eleven":"twelve"},"on_http_response":"hello"}`),
+			expectedIsLegacy: false,
+		},
+		{
+			name:             "legacy key exists, but not at top level",
+			msg:              []byte(`{"on_tcp_connect":"inbound"}`),
+			expectedIsLegacy: false,
+		},
+		{
+			name:             "message is not valid json",
+			msg:              []byte(`ngrok is All-in-one API gateway, Kubernetes Ingress, DDoS protection, firewall, and global load balancing as a service.`),
+			expectedIsLegacy: false,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
```

Comment:
Nit: It's useful to mark the `t.Run()` cases as `t.Parallel()` (and the top-level test too if you want)